libs/runtime/ebpf_epoch.c

```diff
@@ -324,95 +321,70 @@ ebpf_epoch_terminate()
     EBPF_RETURN_VOID();
 }
 
-ebpf_epoch_state_t*
-ebpf_epoch_enter()
+#pragma warning(push)
+#pragma warning(disable : 28167) // Not sure how to annotate that _ebpf_epoch_raise_to_dispatch_if_needed may raise
+// the IRQL.
+void
+ebpf_epoch_enter(_Out_ ebpf_epoch_state_t* epoch_state)
 {
-    uint32_t current_cpu;
-    ebpf_epoch_state_t* epoch_entry = NULL;
-    // Capture preemptible state outside lock
-    bool is_preemptible = ebpf_is_preemptible();
-    current_cpu = ebpf_get_current_cpu();
-
-    if (is_preemptible) {
-        // Prevent APCs from running.
-        ebpf_enter_critical_region();
-
-        // Block until a thread entry is available.
-        ebpf_semaphore_wait(_ebpf_epoch_cpu_table[current_cpu].epoch_table_semaphore);
-        // After the semaphore is acquired, then there is at least 1 available epoch entry.
-    }
-
-    // If the current thread is not preemptible, then there is at least 1 available epoch entry.
-
-    // Grab the CPU lock.
-    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
+    epoch_state->irql_at_enter = _ebpf_epoch_raise_to_dispatch_if_needed();
+    epoch_state->cpu_id = ebpf_get_current_cpu();
 
-    // Get the next available epoch entry.
-    epoch_entry = _ebpf_epoch_next_available_epoch_entry(&_ebpf_epoch_cpu_table[current_cpu]);
-    ebpf_assert(epoch_entry != NULL);
-    _Analysis_assume_(epoch_entry != NULL);
-    ebpf_assert(epoch_entry->epoch == 0);
+    ebpf_epoch_cpu_entry_t* cpu_entry = &_ebpf_epoch_cpu_table[epoch_state->cpu_id];
+    epoch_state->epoch = cpu_entry->current_epoch;
+    ebpf_list_insert_tail(&cpu_entry->epoch_state_list, &epoch_state->epoch_list_entry);
 
-    // Capture the current epoch.
-    epoch_entry->epoch = _ebpf_current_epoch;
-    epoch_entry->is_preemptible = is_preemptible;
-
-    // Release the CPU lock.
-    ebpf_lock_unlock(&_ebpf_epoch_cpu_table[current_cpu].lock, state);
-    return epoch_entry;
+    _ebpf_epoch_lower_to_previous_irql(epoch_state->irql_at_enter);
 }
+#pragma warning(pop)
 
+#pragma warning(push)
+#pragma warning(disable : 28167) // Not sure how to annotate that _ebpf_epoch_raise_to_dispatch_if_needed may raise
+// the IRQL.
 void
```

Comment:
Since ebpf_epoch_exit raises IRQL and restores the original, should it have `_IRQL_requries_same_` annotation here to prevent adding any code paths that don't restore irql?

Reply:
Maybe add _IRQL_requires_same_ and disable any other resulting warning for now?